src/context.rs

```diff
@@ -32,13 +32,42 @@ fn parse_json_visitor<'a>(path_stack: &mut VecDeque<&'a str>, path: &'a str) {
                 continue;
             }
             ".." => {
-                path_stack.pop_back();
+                let item = path_stack.pop_back();
+                if let Some(item) = item {
+                    if let (Some(a), Some(z)) = (item.chars().nth(0), item.chars().last()) {
+                        // if the item we just popped off is an array index,
+                        // pop off one more to make it behave more like the
+                        // Javascript implementation
+                        if a == '[' && z == ']' {
+                            path_stack.pop_back();
+                        }
+                    }
+                }
             }
             _ => {
                 for dot_p in p.split('.') {
-                    // if dot_p != "this" {
-                    path_stack.push_back(dot_p);
-                    // }
+                    let skip = match (path_stack.front(), path_stack.back()) {
+                        (Some(ref a), Some(ref z)) => {
+                            // if this path starts with "this", then we're not going
+                            // to skip anything
+                            if **a == "this" {
+                                false
+                            } else {
+                                // if the path doesn't start with "this" and
+                                // the last element on the deque is "p",
+                                // then don't push it
+                                **z == dot_p
```

Comment:
If the JSON object has same key name in two levels, then there will be no chance to visit the inner one. For example:

```
{
  "a": {
    "a": 99,
    "c": [
      { "d": 100},
      { "d": 200}
    ]
  }
}
```

The second `a` in `../a.a` will be ignored.